packages/compiler/src/render3/r3_ast.ts

```
@@ -43,12 +43,17 @@ export class BoundAttribute implements Node {
   constructor(
       public name: string, public type: BindingType, public securityContext: SecurityContext,
       public value: AST, public unit: string|null, public sourceSpan: ParseSourceSpan,
-      public valueSpan?: ParseSourceSpan, public i18n?: I18nMeta) {}
+      readonly keySpan: ParseSourceSpan, public valueSpan: ParseSourceSpan|undefined,
+      public i18n: I18nMeta|undefined) {}
 
-  static fromBoundElementProperty(prop: BoundElementProperty, i18n?: I18nMeta) {
+  static fromBoundElementProperty(prop: BoundElementProperty, i18n?: I18nMeta): BoundAttribute
+      |null {
+    if (prop.keySpan === undefined) {
```

If keySpan is undefined here, we should just throw an error, because it indicates a bug in our implementation, not user code.

On a separate note, doesn't this mean we'll return null for `BoundElementProperty` for host binding?
I think the solution we discussed is to use sourceSpan as the keySpan for this edge case.
If possible, please look into how we could add a test for this scenario.

---

Yes, this would return null (actually error now after the change) for host bindings. However, `BoundElementProperty`s of host bindings aren't converted to a `BoundAttribute`s but rather converted to a host binding function here: https://github.com/angular/angular/blob/88d7bb838607622c48488a828286c54fe091d8f0/packages/compiler/src/render3/view/compiler.ts#L546. I've added a lengthy comment to `binding_parser` about using `sourceSpan` as the `keySpan` for host bindings.